code/datums/ai/carp/carp_behaviors.dm

```diff
@@ -0,0 +1,67 @@
+/datum/ai_behavior/battle_screech/carp
+	screeches = list("gnashes")
+
+/// This behavior involves attacking a target.
+/datum/ai_behavior/carp_attack
+	behavior_flags = AI_BEHAVIOR_REQUIRE_MOVEMENT | AI_BEHAVIOR_MOVE_AND_PERFORM
+	required_distance = 1
+
+/datum/ai_behavior/carp_attack/perform(delta_time, datum/ai_controller/controller)
+	. = ..()
+	var/mob/living/living_pawn = controller.pawn
+	if(!istype(living_pawn) || !isturf(living_pawn.loc))
+		return
+
+	var/datum/weakref/attack_ref = controller.blackboard[BB_CARP_ATTACK_TARGET]
+	var/atom/movable/attack_target = attack_ref?.resolve()
+	if(!attack_target || !can_see(living_pawn, attack_target, length=AI_CARP_VISION_RANGE))
```

Comment:
vision range should be a blackboard key. These behaviors should be generic and this entire behavior would be better off being more generic as it has no real reason to be only for carps as far as I can tell